packages/cosmic-swingset/src/export-kernel-db.js

```diff
@@ -0,0 +1,405 @@
+    .reduceRight((prev, fn) => prev.finally(fn), exportDone)
+    .catch(err => {
+      startedKit.reject(err);
+      throw err;
+    });
+
+  return {
+    getBlockHeight: () => blockHeight || savedBlockHeight,
+    onStarted: async () => startedKit.promise,
+    onDone: async () => done,
+    stop: async () => {
+      stopped ||= new Error(`Interrupted`);
+      return done.catch(err => {
+        if (err !== stopped) {
+          throw err;
+        }
+      });
+    },
+  };
+};
+
+/**
+ * @typedef {{type: 'started', blockHeight: number}} ExportMessageStarted
+ * @typedef {{type: 'done', error?: Error}} ExportMessageDone
+ * @typedef {ExportMessageStarted | ExportMessageDone} ExportMessage
+ */
+
+/**
+ * @param {string[]} args
+ * @param {object} powers
+ * @param {Partial<Record<string, string>>} powers.env
+ * @param {string} powers.homedir
+ * @param {((msg: ExportMessage) => void) | null} powers.send
+ * @param {Console} powers.console
+ * @param {import('fs/promises')} powers.fs
+ * @param {import('path')['resolve']} powers.pathResolve
+ */
+export const main = async (
+  args,
+  { env, homedir, send, console, fs, pathResolve },
+) => {
+  const processValue = makeProcessValue({ env, args });
+
+  const stateDir =
+    processValue.getFlag('state-dir') ||
+    // We try to find the actual cosmos state directory (default=~/.ag-chain-cosmos)
+    `${processValue.getFlag(
+      'home',
+      `${homedir}/.ag-chain-cosmos`,
+    )}/data/ag-cosmos-chain-state`;
+
+  const stateDirStat = await fs.stat(stateDir);
+  if (!stateDirStat.isDirectory()) {
+    throw new Error('state-dir must be an exiting directory');
+  }
+
+  const exportDir = pathResolve(
+    /** @type {string} */ (processValue.getFlag('export-dir', '.')),
+  );
+
+  const includeExportData = processValue.getBoolean({
+    flagName: 'include-export-data',
+  });
+  const exportMode = processValue.getFlag('export-mode');
+  checkExportMode(exportMode);
+
+  const checkBlockHeight = processValue.getInteger({
+    flagName: 'check-block-height',
+  });
+
+  const verbose = processValue.getBoolean({
+    flagName: 'verbose',
+  });
+
+  const { registerShutdown } = makeShutdown(false);
+
+  const exporter = initiateSwingStoreExport(
+    {
+      stateDir,
+      exportDir,
+      blockHeight: checkBlockHeight,
+      exportMode,
+      includeExportData,
+    },
+    {
+      fs,
+      pathResolve,
+      log: verbose ? console.log : undefined,
+    },
+  );
+
+  registerShutdown(() => exporter.stop());
+
+  exporter.onStarted().then(
+    () =>
+      send?.({
+        type: 'started',
+        blockHeight: /** @type {number} */ (exporter.getBlockHeight()),
+      }),
+    () => {},
+  );
+
+  await exporter.onDone().then(
+    () => send?.({ type: 'done' }),
+    error => {
+      if (send) {
+        send({ type: 'done', error });
+        verbose && console.error('state-sync export failed', error);
+      } else {
+        throw error;
+      }
+    },
+  );
```

Comment:
`send` and `verbose` could be more independent, even if we're paranoid about console.error itself throwing.
```suggestion
  await exporter.onDone().then(
    () => send?.({ type: 'done' }),
    error => {
      try {
        verbose && console.error('state-sync export failed', error);
      } catch(err) {}
      if (send) {
        send({ type: 'done', error });
      } else {
        throw error;
      }
    },
  );
```